### PR TITLE
Rake task and service for publishing a course

### PR DESCRIPTION
**Pull request by inulty-dfe**
## Context

We sometimes need to publish a course through the console. The main scenario is for when the course is in a previous recruitment cycle.
It's good to have a stable service to run this in a repeatable and safe way

## Changes proposed in this pull request

Extract a Service `Courses::PublishService` from the methods in the `Course` model.

The service accepts a user and a course
The task accepts a `uuid` and user email

- I created a Rake tasks which accepts an email address so we don't commit an admins email address to the code.
- I chose the uuid because courses' uuid are unique across all RecruitmentCycles.

I've used the service in the `Publish::CoursesController#publish` action

### CODE REVIEW

**Comment by inulty-dfe** [comment]
[Register request to publish course from 2024 cycle](https://trello.com/c/wYjP8CsD/237-register-request-to-publish-course-from-2024-cycle)

File: app/services/courses/publish_service.rb [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    UnpublishableError = Class.new(StandardError)
+
+    class << self
+      def publish(uuid:, user:)
+        Rails.logger.tagged('Course::PublishService') do |l|
+          l.info "publishing course uuid: #{uuid}"
+        end
+        course = Course.find_by(uuid:)
+
+        raise UnpublishableError unless course.publishable?
```
Comment (inulty-dfe) [inline app/services/courses/publish_service.rb:14]: I don't think it should raise an error. I want to use this in the controller so I don't want to catch this error.

Should this
1) return false if the course is unpulishable?
2) return the course and expect the caller to check if the course was published?

Comment (tomas-stefano) [reply app/services/courses/publish_service.rb:14]: I'd go for the solution 2.

This is how I'd write this:

```ruby
module Courses
  class CoursePublisher
    def initialize(course:, user:)
      @course = course
      @user = user
    end

    def call
      return false unless @course.publishable?

      publish_course
      notify_course_published

      @course
    end

    private

    def publish_course
      Course.transaction do
        @course.publish_sites
        @course.publish_enrichment(@user)
        @course.application_status_open!
      end
    end

    def notify_course_published
      NotificationService::CoursePublished.call(course: @course)
    end
  end
end
```

And on controller:

```ruby
def publish
  course = Course.find_by(uuid: params[:uuid])
  
  if result = Courses::CoursePublisher.new(course: course, user: current_user).call
    redirect_to course_path(result), notice: 'Course published successfully!'
  else
    redirect_to courses_path, alert: 'Course could not be published.'
  end
end
```

I put the notification out of the transaction. Not sure if it should though. 

Also Publish has this pattern of `include ServicePattern`. Not sure if I like to suffix everything with `Service` though.

Comment (inulty-dfe) [reply app/services/courses/publish_service.rb]: Nice, I'll incorporate this into the PR

**Review by tomas-stefano — APPROVED** [review]
🚀 

What do you think to add a documentation or small content on support playbook?

**Pull request merged**